pkg/credentialprovider/apis/criauth/types.go

```diff
@@ -0,0 +1,117 @@
+// plugins that provide container registry credentials.
+type CRIAuthPlugin struct {
+	// name is the name of the credential plugin.  It must match the name of the
+	// executable as seen by the kubelet.
+	Name string
+
+	// imageMatchRefs contains information about the matching image URLs this plugin should
+	// handle along with the cache duration of credentials for any matching images.
+	// Credentials are cached independently per imageMatchRef.
+	ImageMatchRefs []ImageMatchRef
+
+	// ExtraArgs specifies arguments to be passed to the plugin via environment variables.
+	// +optional
+	ExtraArgs []PluginArg
+}
+
+// ImageMatchRef holds information for what image URLs a plugin should handle.
+// The plugin will be called by a kubelet if an image requiring credentials matches
+// the information in imageMatchRef. It also indicates how long the auth plugin
+// should cache credentials in-memory.
+type ImageMatchRef struct {
+	// matchURL is a list of strings used to match against the image property
+	// (sometimes called "registry path") to determine which images to provide
+	// credentials for.  If one of the strings matches the image property, then the
+	// CRIAuthPlugin will be used by kubelet to provide credentials
+	// for the image pull.
+	//
+	// The image property of a container supports the same syntax as the docker
+	// command does, including private registries and tags. A registry path is
+	// similar to a URL, but does not contain a protocol specifier (https://).
+	//
+	// Each matchURL string is a pattern which can optionally contain
+	// a port and a path, similar to the image spec.  Globs can be used in the
+	// hostname (but not the port or the path).
+	//
+	// Globs are supported as subdomains (*.k8s.io) or (k8s.*.io), and
+	// top-level-domains (k8s.*).  Matching partial subdomains is also supported
+	// (app*.k8s.io).  Each glob can only match a single subdomain segment, so
+	// *.io does not match *.k8s.io.
+	//
+	// The image property matches when it has the same number of parts as the
+	// ImageMatcher string, and each part matches.  Additionally the path of
+	// ImageMatcher must be a prefix of the target URL. If the ImageMatcher
+	// contains a port, then the port must match as well.
+	MatchURL string
+	// cacheDuration is the duration the plugin will cache credentials in-memory
+	// for images that matches this reference as indicated by the matchURL field.
+	CacheDuration metav1.Duration
+}
+
+// PluginArg is used for passing extra arguments to the exec credential plugin binary.
+// Exec plugins expect the arguments to be passed in the format "key=value".
+type PluginArg struct {
+	Name  string
+	Value string
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// CRIAuthPluginRequest includes the image that the kubelet requires authenication for.
+// Kubelet will pass this request object to the plugin via stdin.
+type CRIAuthPluginRequest struct {
```

Comment:
Can single plugin be called in parallel? Would there be any benefit from passing correlation ID to match logs on case of error troubleshooting?

Reply:
> Can single plugin be called in parallel? Would there be any benefit from passing correlation ID to match logs on case of error troubleshooting?

I think so but likely not something we want to include in the alpha version of this since we are going to be caching credential based on the match URL anyways